feat: unmanaged installs

mistydemeo opened the pull request:
The new "unmanaged" installation mode allows for something close to "untar and move on" installations. It disables modifying the `PATH` and creation of environment files, and ensures all files are installed into a single flat directory no matter what other configuration is in place. It also disables installing a standalone updater and the creation of an install receipt.

The new "don't install an updater/receipt" behaviour is also controllable behind a new public environment variable, `$APPNAME_DISABLE_UPDATE`.

Fixes #1400.

ashleygwilliams commented:
q: does this still attempt receipt creation? i think it'd be best to just not try at all

mistydemeo commented:
The issue sort of identified a conflict between the two user groups this attempts to serve - one would care about receipts, the other is neutral on them. I tried to split the difference by *attempting* receipt creation, but treating receipt creation failure as harmless.

ashleygwilliams commented:
hrm, i think splitting the difference here is maybe the less good option and we should consider 2 modes

ashleygwilliams commented:
part of the reason i say this is that an unmanaged install is desirable for ephemeral environments which are usually automated- receipt creation is an i/o operation and while not particularly expensive it's not free- this feature in my mind is specifically for the "conductor" persona and they are the type that care about milliseconds

mistydemeo commented:
Based on our previous discussion, I've made a few changes:

* We no longer split the difference between two groups of users; this is now completely focused on automated/CI usecases, and the updater/receipt is unconditionally disabled.
* The updater disabling functionality can be enabled separately with a new environment variable.
* The environment variables have been moved into a new grouping to make reasoning about them as a group easier:

```diff
-      "install_dir_env_var": "CARGO_DIST_SCHEMA_INSTALL_DIR",
+      "environment_variables": {
+        "install_dir_env_var": "CARGO_DIST_SCHEMA_INSTALL_DIR",
+        "unmanaged_dir_env_var": "CARGO_DIST_SCHEMA_UNMANAGED_INSTALL",
+        "disable_update_env_var": "CARGO_DIST_SCHEMA_DISABLE_UPDATE"
+      },
```

ashleygwilliams commented:
we could probably name that `env` for brevity's sake

mistydemeo commented:
Sounds good, done.

```diff
-      "environment_variables": {
+      "env": {
```

Gankra commented on cargo-dist/templates/installer/installer.ps1.j2:
```diff
@@ -55,6 +55,18 @@ $receipt = @"
+
+$unmanaged_install = $env:{{ unmanaged_dir_env_var }}
+if ($unmanaged_install) {
+  $NoModifyPath = $true
```
We should be deleting the -NoModifyPath `param`, people keep finding these weird inconsistent APIs we just shouldn't support them

Gankra commented on cargo-dist-schema/src/lib.rs:
```diff
@@ -610,12 +621,21 @@ impl DistManifest {
+            let env_app_name = name.to_ascii_uppercase().replace('-', "_");
+            let install_dir_env_var = format!("{env_app_name}_INSTALL_DIR");
+            let unmanaged_dir_env_var = format!("{env_app_name}_UNMANAGED_INSTALL");
+            let disable_update_env_var = format!("{env_app_name}_DISABLE_UPDATE");
```
not DISABLE_UPDATER?

ashleygwilliams replied:
my thought here was that the user cares about the "update" behavior. there is not always an "updater" so "update" focused on the behavior the user understands, not the impl

Gankra commented on cargo-dist-schema/src/lib.rs:
```diff
@@ -610,12 +621,21 @@ impl DistManifest {
-            let install_dir_env_var =
-                Some(name.to_ascii_uppercase().replace('-', "_") + "_INSTALL_DIR");
+            let env_app_name = name.to_ascii_uppercase().replace('-', "_");
+            let install_dir_env_var = format!("{env_app_name}_INSTALL_DIR");
+            let unmanaged_dir_env_var = format!("{env_app_name}_UNMANAGED_INSTALL");
+            let disable_update_env_var = format!("{env_app_name}_DISABLE_UPDATE");
```
Should have *_NO_MODIFY_PATH in here too

Gankra commented on cargo-dist-schema/src/lib.rs:
```diff
@@ -610,12 +621,21 @@ impl DistManifest {
+            let install_dir_env_var = format!("{env_app_name}_INSTALL_DIR");
+            let unmanaged_dir_env_var = format!("{env_app_name}_UNMANAGED_INSTALL");
+            let disable_update_env_var = format!("{env_app_name}_DISABLE_UPDATE");
+
```
We should maybe also expose FLAT_INSTALL_LAYOUT as a raw option too? less certain of that

mistydemeo replied:
How about we circle back around to publicizing this around the time we also give developers full control over the install layout?

Gankra commented on cargo-dist/templates/installer/installer.ps1.j2:
```diff
@@ -55,6 +55,18 @@ $receipt = @"
 "@
 $receipt_home = "${env:LOCALAPPDATA}\{{ app_name }}"
 
+if ($env:{{ disable_update_env_var }}) {
```
We need to be checking INSTALLER_INSTALL_UPDATE(R) as well here.

Gankra replied:
disregard this, we are explicitly *not* adding generic INSTALLER_ vars in this PR

Gankra commented on cargo-dist/templates/installer/installer.sh.j2:
```diff
@@ -22,6 +22,19 @@ ARTIFACT_DOWNLOAD_URL="${INSTALLER_DOWNLOAD_URL:-{{ base_url }}}"
 PRINT_VERBOSE=${INSTALLER_PRINT_VERBOSE:-0}
 PRINT_QUIET=${INSTALLER_PRINT_QUIET:-0}
 NO_MODIFY_PATH=${INSTALLER_NO_MODIFY_PATH:-0}
+ALLOW_RECEIPTS_TO_FAIL=0
```
I don't think we need/want to add this -- certainly the powershell installer doesn't so this is just making it more inconsistent?

mistydemeo replied:
Yeah, this isn't properly used anymore so I'll take it out.

Gankra commented on cargo-dist/templates/installer/installer.sh.j2:
```diff
@@ -22,6 +22,19 @@ ARTIFACT_DOWNLOAD_URL="${INSTALLER_DOWNLOAD_URL:-{{ base_url }}}"
 PRINT_QUIET=${INSTALLER_PRINT_QUIET:-0}
 NO_MODIFY_PATH=${INSTALLER_NO_MODIFY_PATH:-0}
+ALLOW_RECEIPTS_TO_FAIL=0
+if [ "{{ '${' }}{{ disable_update_env_var }}:-0}" = "1" ]; then
```
need to be checking an INSTALLER_ variant as well

Gankra commented on cargo-dist/templates/installer/installer.sh.j2:
```diff
@@ -22,6 +22,19 @@ ARTIFACT_DOWNLOAD_URL="${INSTALLER_DOWNLOAD_URL:-{{ base_url }}}"
+else
+    INSTALL_UPDATER=1
+fi
+UNMANAGED_INSTALL="{{ '${' }}{{ unmanaged_dir_env_var }}:-}"
```
need to be checking an INSTALLER_ variant as well

Gankra commented on cargo-dist/templates/installer/installer.sh.j2:
```diff
@@ -241,12 +254,23 @@ download_binary_and_run_installer() {
+                say "this installation will not be able to self-update"
+                local _retval=0
+            else
+                err "unable to create receipt directory at $RECEIPT_HOME"
```
"install receipt"

mistydemeo commented on cargo-dist/templates/installer/installer.ps1.j2:
```diff
@@ -267,13 +285,13 @@ function Invoke-Installer($artifacts, $platforms) {
-    $dest_dir = $force_install_dir
-    $dest_dir_lib = $dest_dir
-{%- endif %}
+    if (!$flat_install_layout) {
```
Wow yeah, good point. Well, I'll swap to `-not` for consistency

Gankra commented on cargo-dist/templates/installer/installer.ps1.j2:
```diff
@@ -55,6 +50,24 @@ $receipt = @"
+  $install_updater = $true
+}
+
+$NoModifyPath = $false
```
format of this variable name is weird now but not a big deal?

mistydemeo replied:
Marking as resolved since I've decided to keep the param for consistency with shell.

Gankra commented on cargo-dist/templates/installer/installer.ps1.j2:
```diff
@@ -55,6 +50,24 @@ $receipt = @"
+
+$unmanaged_install = $env:{{ unmanaged_dir_env_var }}
+
+if ($unmanaged_install) {
+  $NoModifyPath = $true
+  $install_updater = $false
+}
```
I think the "mode" variable should get applied first, and then the other overrides should kick in..?

Gankra commented on cargo-dist/templates/installer/installer.sh.j2:
```diff
@@ -21,7 +21,22 @@ APP_VERSION="{{ app_version }}"
+    INSTALL_UPDATER=1
+fi
+UNMANAGED_INSTALL="{{ '${' }}{{ unmanaged_dir_env_var }}:-}"
+if [ -n "${UNMANAGED_INSTALL}" ]; then
+    NO_MODIFY_PATH=1
+    INSTALL_UPDATER=0
+fi
```
ditto on mode should go first..?

mistydemeo replied:
Honestly I think "unmanaged mode" is opinionated enough that it probably should override the individual options instead of the other way around. I see the individual options as being useful in normal mode, not for this.

Gankra reviewed (approved):
think we're basically there, a few more nits

mistydemeo commented:
I just realized that the shell installer *does* provide a CLI param for `--no-modify-path`, so I've removed the change where I removed that feature from powershell. Instead, I've just added the env override while retaining CLI support.
